packages/cli/src/Version.ts

```diff
@@ -74,29 +76,47 @@ export class Version implements Command {
 
     const platform = await getPlatform()
     const cliQueryEngineBinaryType = getCliQueryEngineBinaryType()
-    const introspectionEngine = await this.resolveEngine(BinaryType.introspectionEngine)
-    const migrationEngine = await this.resolveEngine(BinaryType.migrationEngine)
-    // TODO This conditional does not really belong here, CLI should be able to tell you which engine it is _actually_ using
-    const queryEngine = await this.resolveEngine(cliQueryEngineBinaryType)
-    const fmtBinary = await this.resolveEngine(BinaryType.prismaFmt)
+
+    const [enginesMetaInfo, enginesMetaInfoErrors] = await getEnginesMetaInfo()
+
+    const enginesRows = enginesMetaInfo.map((engineMetaInfo) => {
+      return match(engineMetaInfo)
+        .with({ 'query-engine': P.select() }, (currEngineInfo) => {
+          return [
+            `Query Engine${cliQueryEngineBinaryType === BinaryType.libqueryEngine ? ' (Node-API)' : ' (Binary)'}`,
+            currEngineInfo,
+          ]
+        })
+        .with({ 'migration-engine': P.select() }, (currEngineInfo) => {
+          return ['Migration Engine', currEngineInfo]
+        })
+        .with({ 'introspection-engine': P.select() }, (currEngineInfo) => {
+          return ['Introspection Engine', currEngineInfo]
+        })
+        .with({ 'format-binary': P.select() }, (currEngineInfo) => {
+          return ['Format Binary', currEngineInfo]
+        })
+        .exhaustive()
+    })
 
     const prismaClientVersion = await getInstalledPrismaClientVersion()
 
     const rows = [
       [packageJson.name, packageJson.version],
       ['@prisma/client', prismaClientVersion ?? 'Not found'],
       ['Current platform', platform],
-      [
-        `Query Engine${cliQueryEngineBinaryType === BinaryType.libqueryEngine ? ' (Node-API)' : ' (Binary)'}`,
-        this.printBinaryInfo(queryEngine),
-      ],
-      ['Migration Engine', this.printBinaryInfo(migrationEngine)],
-      ['Introspection Engine', this.printBinaryInfo(introspectionEngine)],
-      ['Format Binary', this.printBinaryInfo(fmtBinary)],
+
+      ...enginesRows,
+
       ['Default Engines Hash', packageJson.dependencies['@prisma/engines'].split('.').pop()],
       ['Studio', packageJson.devDependencies['@prisma/studio-server']],
     ]
 
+    if (enginesMetaInfoErrors.length > 0) {
+      process.exitCode = 1
+      enginesMetaInfoErrors.forEach((e) => console.error(e))
+    }
```

Comment:
Add a comment here? it's a bit obscure if one doesn't know

Reply:
```suggestion
    /**
     * If reading Rust engines metainfo (like their git hash) failed, display the errors to stderr,
     * and let Node.js exit naturally, but with error code 1.
     */
    if (enginesMetaInfoErrors.length > 0) {
      process.exitCode = 1
      enginesMetaInfoErrors.forEach((e) => console.error(e))
    }
```